### ISSUE TITLE
Upgrade to support v3 language (e.g. local functions)

### ISSUE THREAD

**Issue by conficient** [toddams/RazorLight#307]
The current RazorLight (excellent library, thank you!) supports the Razor 2.2 language syntax. 

With ASP.NET Core 3.0 an updated language syntax and base library 3.0.0 and 3.1.0 have been introduced with `Microsoft.AspNetCore.Razor.Language` - is there any plan to upgrade RazorLight to support this?

Example of new language feature: local functions (a `@Helper` replacement)
```
<div>
@Hello()
</div>

@functions 
{
   void Hello() 
   {
       <p>Hello world</p>
   }
}
```

**Comment by jzabroski** [comment]
tl;dr: If you want to accept a PR, I'll review, but please include tests.  Thank you.

Long answer:

> The current RazorLight (excellent library, thank you!) supports the Razor 2.2 language syntax.

How did you determine it supports the Razor 2.2 language syntax?

It's called RazorLight for a reason: It's a sub-set of Razor.  There are likely features missing, even from the Razor 2.2 language syntax.

> is there any plan to upgrade RazorLight to support this?

It's an open source library and not a business - nothing is "planned".  The original author, Ivan Balan, has been working on other projects while I help maintain the library.  I'm happy to accept PRs, but _I personally don't have any use for local functions in my projects, _and never will, because they're not unit testable_.  I'd rather have static methods on a static class instead, because I can unit test those.

**Comment by conficient** [comment]
Thanks for the reply - yes I'll try to submit a PR for the upgrade.
I tried to create a Razor view with a `@functions` block and a method that contained HTML (like the old '@helper' syntax) and the template compilation step failed, so this is why I assumed it only supported 2.2 syntax. The 'helper' method was introduced in .NET Core 3.0

**Comment by jzabroski** [comment]
The way Razor works internally, and it's changed a bit from .net core version to version, is it's basically a set of plug-ins called directives (ugh - hate the angular-like vocabulary) and these directives help process "chunks" of text.  Dave Glick, who maintains Wyam.io (a great static site generator), has written some blog posts about the internals: https://daveaglick.com/posts/the-bleeding-edge-of-razor

**Comment by toddams** [comment]
Have you had a chance to create a PR, @conficient? :)

**Comment by conficient** [comment]
Sorry work commitments since everyone came back has put it on the back burner. Feel free to close this issue if you wish.

**Comment by rmannibucau** [comment]
any hope it becomes a feature, it would avoid to need to create a project for simple helpers like date formatting, character separated list etc

**Comment by conficient** [comment]
Sorry I no longer use RazorLight. I now use a library I wrote called `BlazorTemplater` that uses Blazor components for rendering HTML as this is a much better system than old Razor views since Blazor components are compiled as classes. Since .NET 8 the framework now also supports this approach, but my library works for .NET Core 3.1 upward. Welcome to try it out at https://github.com/conficient/BlazorTemplater

**Comment by rmannibucau** [comment]
sadly blazor does not match much decoupling view and model compilation units (ie the project defining view is a downstream project of the model/controller one and does not have any explicit csproj/compilation unit compared cshtml) so I can't move to blazor flavor :(

**Comment by conficient** [comment]
Not quite sure I understand. The BlazorTemplater library can live in a class library or in a web project. It also doesn't require the use of a Blazor web application to work. The library has an example using a console app, for example. 

Passing a model to a Blazor component is easy (see the documentation on how to do this - you just set a property). 

Let me know if I can help.

**Comment by rmannibucau** [comment]
@conficient I'm using it for a kind of static site generator so the project does not have .NET projec (no csproj or alike).
It is installed as a dotnet tool then ran from the CLI.
All the user does is to create some model (.adoc), some views (.cshtml but .razor would be okish) and done: https://rmannibucau.github.io/SBSharp/how-it-works.html#_your_first_content.

Originally I was tempted to reimplement a JSP like solution but since razor was kind of matching for .NET I just went with razorlight but I miss helpers/inline functions to make it fully usable - happy to evalute blazor alternative but as of today I'm not sure how it would work.

**Comment by conficient** [comment]
Hmm, yes I don't think BlazorTemplater could help here. It may be possible to compile `.razor` files without a project but it's not something I know how to do.

It is possible to generate a static site using BlazorTemplator but only where a project exists to compile the `.razor` files before rendering, not as a `dotnet` tool.
